all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+There are many ways to access Temporal programmatically including CLI, SDK, and other components that leverage those
+such as samples and Terraform providers.
+
+Temporal needs a way for users to provide client configuration without hardcoding in CLI arguments or SDK options. This
+will allow code to seamlessly switch between profiles such as between a dev-server and cloud. The same configuration
+must work across all SDKs/CLIs.
+
+### Why?
+
+What are the reasons for building this?
+
+* Users want to switch environments without changing their code, but today they cannot.
+* Our samples should be able to be run on multiple environments including cloud, but today they are localhost only.
+
+### Goals
+
+* Common, well-specified configuration format usable by Temporal tooling for client options
+* Implementation in all Temporal SDKs for reading and writing of this configuration
+* All Temporal CLI/SDKs updated to have easy client creation using this configuration
+* Common on-disk location in every platform for reading/writing this format
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
+* Use external configuration by default. Users need to opt-in to this for compatibility and clarity reasons (but it is
+  of course easy).
+
+### High-level Usage Examples
+
+#### Simple Switch from Local to Cloud
+
+Similar to `temporal env` today, you can use the CLI to manage configurations with `temporal config`.
+
+For example, say you had this Go code for starting a workflow:
+
+```go
+options, err := client.LoadOptionsFromConfig("default")
+if err != nil {
+  return err
+}
+cl, err := client.Dial(options)
+run, err := cl.ExecuteWorkflow(
+  ctx,
+  client.StartWorkflowOptions{ID: "my-id", TaskQueue: "my-task-queue"},
+  MyWorkflow,
+)
+```
+
+By default this only works on `localhost:7233` with namespace `default` when you run it. But if you were a cloud user,
+you could do this in environment variables for mTLS auth:
+
+    export TEMPORAL_ADDRESS=my-ns.a1b2c.tmprl.cloud:7233
+    export TEMPORAL_NAMESPACE=my-ns.a1b2c
+    export TEMPORAL_TLS_CLIENT_CERT_PATH=path/to/my/client.pem
+    export TEMPORAL_TLS_CLIENT_KEY_PATH=path/to/my/client.key
```

Comment:
I'm confused how these relate to "default"

Reply:
Could users get sniped by the default lack of isolation by using these bash commands or the commands below?  (I know I personally have.)  Would rather see a file whose impacts are scoped by default rather than the equivalent of global variables.

Reply:
> I'm confused how these relate to "default"

As @josh-berry requested, I will add a section clarifying how configuration is read from environment variables and the file. In this case, `TEMPORAL_ADDRESS` is the env var equivalent of `TEMPORAL_DEFAULT_ADDRESS` which is the equivalent of setting the `address` for profile `default` which is the equivalent of providing `--address` on the CLI or setting it in the SDK client options.

> Could users get sniped by the default lack of isolation by using these bash commands or the commands below? Would rather see a file whose impacts are scoped by default rather than the equivalent of global variables.

This is your choice as a user, and we even allow env vars to be disabled when loading in an SDK, but we can't just eschew environment variables for this use case because people may set them accidentally.